### ISSUE TITLE
Feature request: create variable to get copy mode's scrollback line number

### ISSUE THREAD

**Issue by ericpruitt** [tmux/tmux#3787]
### Issue description

I'm trying to correlate the position of the cursor in copy mode with output generated by capture-pane, but I don't see a way to get the position of the cursor in the history in any of the "copy_cursor_*" variables. I thought "copy_cursor_line" was what I wanted, but that just returns the string which is only useful if the string is unique.

### Required information

```
$ uname -sp && tmux -V && echo $TERM
Linux unknown
tmux next-3.4
tmux-256color
```

**Comment by nicm** [comment]
You should be able to use `copy_cursor_y` and `scroll_position`?

**Comment by ericpruitt** [comment]
I don't think that works unless I'm misunderstanding your suggestion. The numbers get incremented for each displayed line, so when a line is wider than the pane, the values no longer have a 1:1 mapping with the lines output of capture-pane. For the same reason, copy_cursor_x can't be used to figure out which column the cursor is on if the line is wider than the pane.

**Comment by ericpruitt** [comment]
Actually, I think that might work after all. I thought the output of capture-pane is different from what it actually is. I'll try to write I'm trying to create and follow-up afterward.

**Comment by nicm** [comment]
Oh yeah there is probably no way to do this if you use `-J` because copy mode does not track the position on joined lines (it has no need to).

**Comment by ericpruitt** [comment]
Yeah, you're right, I can't do it with "-J". I would prefer to have a feature that does with with "-J" and think supporting these variables would be useful, but if I leave off "-J", using copy_cursor_x and scroll_position is a mostly tolerable compromise. I'll let you decide whether or not you want this closed or if you plan on exposing the necessary variables.

**Comment by ericpruitt** [comment]
I ran into another limitation of copy_cursor_x: since it uses the physical position of the cursor, regardless of whether "-J" is used, the only way to correlate the position when the output contains wide characters requires the use of something like wcwidth or utf8proc. That is not really feasible for a lot of common CLI *NIX programs like `sh`, `sed`, (portable) `awk`, etc.

**Comment by nicm** [comment]
What would you like to see here? Having a format which shows cursor position including wrapped lines would be possible but it would need to walk the entire history so it may be slow and you would not want to use it too often.

**Comment by nicm** [comment]
You would probably just need to do something like:
~~~~
ey = py = size - oy + cy;
for (yy = 0; yy < ey; yy++)  {
    gl = grid_get_line(yy);
    if (gl->flags & GRID_FLAG_WRAP)
        py--;
}
return py;

**Comment by ericpruitt** [comment]
Right now, there are two things I'm trying to do, both of which require being able figure out exactly where the cursor is in the output of capture-pane, but I just realized that I don't need to get the x and y position of the cursor simultaneously for my use cases.

The first thing I'm trying to do is to be able to use Less as the scrollback pager in place of tmux's copy mode because Less has the ability to do a number of things that tmux can't out of the box. In my tmux configuration file, I have `bind-key    ]   run-shell "popup -Eo scrollback"` which opens the scrollback in Less. When I'm in copy mode, I want to be able to position Less on the same line as the cursor so the transition from tmux's copy mode to Less is relatively seamless.

The other thing I'm trying to do is be able to write scripts for tmux that allow me to perform operations on the text under the cursor e.g. opening a URL or file that my cursor is on top of. In that case, when I use capture-pane's "-J" option, I just need to be able to get the cursor position in just that line because I can use copy_cursor_line then use the character offset, but that only works if I have the **character** offset. copy_cursor_x in conjunction with copy_cursor_line requires the use of _wcwidth(3)_ or something similar.

Does that make sense?

**Comment by nicm** [comment]
Maybe we could make a flag to `capture-pane` to output some special escape sequence to mark the position of the cursor if the pane is in copy mode.

The other option would be to add a format for the line of the cursor after joined as I said above.

**Comment by ericpruitt** [comment]
Adding a marker to indicate the cursor position would solve the issue of trying to operate on text under the cursor, but it doesn't address the use case of being able to figure out how the view is positioned because the cursor can move around without the viewport changing, but if there was (also) a marker to indicate the top left of the viewport, I think that could work.

**Comment by nicm** [comment]
Did you get something that works here?

**Comment by ericpruitt** [comment]
> Did you get something that works here?

I don't understand this question. Was there a feature added that I missed?

**Comment by nicm** [comment]
I mean, did you find some way of doing what you wanted, or look at implementing one of the things we talked about?

**Comment by ericpruitt** [comment]
No. I don't think there's a sane way to do this without modifying tmux's source code, but I haven't tried to implement changes to address this. It's not an issue I run into particularly often right now because I'm not working, so it hasn't been high on my priority list.

**Comment by nicm** [comment]
Try this please: [copy_join.diff.txt](https://github.com/user-attachments/files/23257847/copy_join.diff.txt)

This adds two format variables when in copy mode:

- `copy_cursor_char_x` - character count from beginning of line, so if you are at the end of `X一X` it should be three rather than four for `copy_cursor_x`.
- `copy_cursor_join_y` - count of joined lines up to cursor line (including history).

**Comment by ericpruitt** [comment]
I was able to apply the patch and will try to use the new variables to implement a couple of scripts I wanted to write in the next day or two.

**Comment by ericpruitt** [comment]
For some reason Ctrl+Shift+Tab doesn't work when I build the current tmux HEAD, so I want to figure out what's going on there before I work on the scripts that use the new format strings.

**Comment by nicm** [comment]
I'm going to close this for the moment, but let me know if these changes work.
